File: p111e/min_depth.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
  if (root == null) return 0;
  let ldepth = minDepth(root.left);
  let rdepth = minDepth(root.right);
  if (ldepth > 0 && rdepth > 0) return 1 + Math.min(ldepth, rdepth);
  else return 1 + ldepth + rdepth;
};
// TESTS
[
  ['1', 1],
  ['1,2,#,#,#', 2],
  ['1,2,#,#,3,#,#', 2],
  ['1,2,3,4,#,#,#,#,#', 4],
  ['1,#,2,#,3,#,4,#,#', 4],
  ['1,#,2,3,4,#,#,#,#', 4],
  ['1,#,2,3,#,#,4,#,#', 3],
  ['3,9,#,#,20,15,#,#,7,#,#', 2],
].forEach(([tree, expected]) => {
  const actual = minDepth(Tree.deserialize(tree));
  console.log('Max depth of', tree, '->', actual);
  console.assert(actual === expected);
});

File: p111e/min_depth.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode


class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        ld = self.minDepth(root.left)
        rd = self.minDepth(root.right)
        if ld > 0 and rd > 0:
            return 1 + min(ld, rd)
        else:
            return 1 + ld + rd


# TESTS
for tree, expected in [
    ("#", 0),
    ("1,#,#", 1),
    ("1,#,2,#,#", 2),
    ("1,2,#,#,3,#,#", 2),
    ("3,9,#,#,20,15,#,#,7,#,#", 2),
    ("2,#,3,#,4,#,5,#,6,#,#", 5),
]:
    sol = Solution()
    actual = sol.minDepth(TreeNode.deserialize(tree))
    print("Max depth of", tree, "->", actual)
    assert actual == expected
